smolyay/surrogate.py

```diff
@@ -243,23 +264,44 @@ def __call__(self, x):
         """
         if self._coefficients is None:
             raise ValueError('Function needs training!')
-        x = numpy.array(x, copy=False, ndmin=1)
-        if x.shape != (self.dimension, ):
-            raise IndexError('Input must match dimension of domain.')
-        if not (numpy.all(x >= self._domain[:, 0])
-                and numpy.all(x <= self._domain[:, 1])):
+        input_shape = numpy.shape(x)
+        x = numpy.array(x, copy=False, ndmin=2)
+        if self.dimension == 1:
+            if len(input_shape) <= 1:
+                # the cast to 2d puts these in wrong order, so transpose
+                x = x.T
+            elif x.shape[-1] > 1:
+                x = x[..., numpy.newaxis]
+        
+        if x.shape[-1] != self.dimension:
+            raise IndexError("Input must match dimension of domain")
+        if self.dimension > 1:
+            oob = any(numpy.any(x[..., i] < self.domain[i][0]) or 
+                    numpy.any(x[..., i] > self.domain[i][1]) 
+                    for i in range(self.dimension))
+        else:
+            oob = (numpy.any(x < self.domain[0]) or 
+                    numpy.any(x > self.domain[1]))
+        if oob:
             raise ValueError('x must lie in domain of surrogate')
         # transform point into basis domain and evaluate
         x_scaled = self._mapdomain(x, self._domain, [[-1, 1]]*self.dimension)
         value = 0
         for coeff, basis in zip(self._coefficients, self.grid.basis_functions):
             if self.dimension > 1:
-                term = numpy.product(
-                    [basis_i(x_i) for basis_i, x_i in zip(basis, x_scaled)])
+                term = numpy.prod(
+                        [basis[i](x_scaled[...,i]) for i in range(len(basis))],
+                        axis=0)
             else:
-                term = basis(x_scaled[0])
+                term = basis(x_scaled)
             value += coeff*term
 
+        if (self.dimension == 1 and input_shape == ()) or (self.dimension > 1 and len(input_shape) == 1):
+            value = value.item()
+        elif (self.dimension == 1 and (len(input_shape) == 1 or input_shape[-1] != 1)):
+            value = numpy.reshape(value, input_shape)
+        else:
+            value = numpy.reshape(value, input_shape[:-1])
```

Comment:
I think this would a little simpler and easier to read?
```suggestion
        if (self.dimension == 1 and len(input_shape) == 0) or (self.dimension > 1 and len(input_shape) == 1):
            value = value.item()
        else:
            if self.dimension == 1 and (len(input_shape) <= 1 or input_shape[-1] > 1):
                # an extra dimension was added in these cases, so use original input shape
                output_shape = input_shape
            else:
                # evaluation of the function drops the last dimension from x
                output_shape = x.shape[:-1]
            value = numpy.reshape(value, output_shape)
```